src/librustc/mir/interpret/value.rs

```diff
@@ -92,277 +86,171 @@ impl<'tcx> ty::TypeFoldable<'tcx> for Value {
+                        defined: pointer_size,
+                    })
             }
-            PrimVal::Ptr(ptr) => Ok(Pointer::from(ptr.wrapping_signed_offset(i, layout))),
-            PrimVal::Undef => err!(ReadUndefBytes),
+            }
+            Scalar::Ptr(ptr) => Ok(Scalar::Ptr(ptr.wrapping_signed_offset(i, layout))),
         }
     }
 
-    pub fn is_null(self) -> EvalResult<'tcx, bool> {
-        match self.primval {
-            PrimVal::Bytes(b) => Ok(b == 0),
-            PrimVal::Ptr(_) => Ok(false),
-            PrimVal::Undef => err!(ReadUndefBytes),
+    pub fn is_null_ptr<C: HasDataLayout>(self, cx: C) -> EvalResult<'tcx, bool> {
+        match self {
+            Scalar::Bits {
+                bits, defined,
+            } => if defined < cx.data_layout().pointer_size.bits() as u8 {
+                err!(ReadUndefBytes)
+            } else {
+                Ok(bits == 0)
+            },
+            Scalar::Ptr(_) => Ok(false),
         }
     }
 
-    pub fn to_value_with_len(self, len: u64) -> Value {
-        Value::ByValPair(self.primval, PrimVal::from_u128(len as u128))
+    pub fn to_value_with_len<C: HasDataLayout>(self, len: u64, cx: C) -> Value {
+        Value::ScalarPair(self, Scalar::Bits {
+            bits: len as u128,
+            defined: cx.data_layout().pointer_size.bits() as u8,
+        })
     }
 
-    pub fn to_value_with_vtable(self, vtable: MemoryPointer) -> Value {
-        Value::ByValPair(self.primval, PrimVal::Ptr(vtable))
+    pub fn to_value_with_vtable(self, vtable: Pointer) -> Value {
+        Value::ScalarPair(self, Scalar::Ptr(vtable))
     }
 
     pub fn to_value(self) -> Value {
-        Value::ByVal(self.primval)
-    }
-}
-
-impl ::std::convert::From<PrimVal> for Pointer {
-    fn from(primval: PrimVal) -> Self {
-        Pointer { primval }
+        Value::Scalar(self)
     }
 }
 
-impl ::std::convert::From<MemoryPointer> for Pointer {
-    fn from(ptr: MemoryPointer) -> Self {
-        PrimVal::Ptr(ptr).into()
+impl From<Pointer> for Scalar {
+    fn from(ptr: Pointer) -> Self {
+        Scalar::Ptr(ptr)
     }
 }
 
-/// A `PrimVal` represents an immediate, primitive value existing outside of a
+/// A `Scalar` represents an immediate, primitive value existing outside of a
 /// `memory::Allocation`. It is in many ways like a small chunk of a `Allocation`, up to 8 bytes in
-/// size. Like a range of bytes in an `Allocation`, a `PrimVal` can either represent the raw bytes
-/// of a simple value, a pointer into another `Allocation`, or be undefined.
+/// size. Like a range of bytes in an `Allocation`, a `Scalar` can either represent the raw bytes
+/// of a simple value or a pointer into another `Allocation`
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, RustcEncodable, RustcDecodable, Hash)]
-pub enum PrimVal {
+pub enum Scalar {
     /// The raw bytes of a simple value.
-    Bytes(u128),
+    Bits {
+        /// number of bits that are valid and may be read
```

Comment:
"Number of bits" is not very useful... when the number is 4, which 4 bits are the ones that I can read?